lib/symbValue.ml

```diff
@@ -341,8 +341,18 @@ module
       | _ -> binop_cs_cs Op.Or Cst.Scalar.logor v1 v2
 
   and xor v1 v2 =
-    if compare v1 v2 = 0 then zero else
-    binop Op.Xor (Cst.Scalar.logxor) v1 v2
+    let do_xor () = binop Op.Xor Cst.Scalar.logxor v1 v2 in
+    match v1,v2 with
+    | (Val (Symbolic id1),Val (Symbolic id2))
+      when Constant.symbol_eq id1 id2
+        -> zero
+    | (Val cv1, _ )
+      when compare v1 v2 = 0
+      ->
+        (match Cst.zero_of cv1 with
+        | Some zero -> Val zero
+        | _ -> do_xor ())
+    | _ -> do_xor ()
```

Comment:
Hi @yuxiliu-arm I do not think we gain much here. Namely, the efficiency gain is huge when `v1` and `v2` are teh same variable (`Var _`). Here we do not gain a lot if anything, because `do_xor` will perform teh reduction to zero anyway. A more efficient change would be to apply the optimisation `if compare v1 v2 = 0 then zero else ...` when the architecture is not `ASL`. Doing so, we assume that integers are the only meaningful scalars, which I guess is the case for all architectures except `ASL`  where scalars also include bitvectors.

Reply:
I see.

> Doing so, we assume that integers are the only meaningful scalars, which I guess is the case for all architectures except ASL where scalars also include bitvectors.

I didn't know that was the case, so I only implemented the conservative change for concrete values. `zero_of` is indeed the way you described (always returns `Some zero` for al backends except ASL). I guess a simpler solution is to just extend `Constant.S` to have `is_xor_result_always_numeric : bool`?